### ISSUE TITLE
[FEATURE]: Clarity on referring to external VEX source from SBOM

### ISSUE THREAD

**Issue by RingoDev** [CycloneDX/specification#548]
## Describe the feature

I would like to be able to strictly typed refer to external VEX information outside of a CDX BOM.

We are already distributing SBOMs with our container images as In-Toto attestations and are very happy that scanners like Trivy are able to pick up the BOM and use it for the most accurate scan results we ever had.

What is missing now is that we would like to be able to communicate vulnerability information (e.g. VEX statements) not hard coded in the BOM itself but rather serve an associated file somewhere else over HTTPS so we are able to make additions when the SBOM was already released to consumers.

 What I was thinking about is to use the "externalReferences" in the "exploitability-statement" type to link to such an external VEX file and trying to get SBOM scanners to adopt resolving such references for immediated assessment of known false positives.

That field seems to be exactly made for that use-cases as being [typed as containing either an **URL** or a **BOM-Link**](https://github.com/CycloneDX/specification/blob/268138039498a4f390ddc5d7e7e78857a951986f/schema/bom-1.6.schema.json#L1710-L1720).

As far as I understand it, the BOM-Link is an abstract link between documents, but would not tell some kind of scanner e.g. where to retrieve it.

The URL type on the other hand could contain any kind of resource on the other end - e.g. a URL to a PDF statement, an HTML page or _maybe_ a CDX BOM with VEX information. This makes it a bit hard to make a case for e2e automation. 

## Possible solutions

What I would be wishing for is 2 things:

* a statement from your side / the community if that is actually a valid usecase that makes sense to be supported by tools interacting with an SBOM for security scanning -> discussion on trivy repo for reference: https://github.com/aquasecurity/trivy/discussions/8047
* some kind of concept where we can provide additionaly typing to the URL to make it easier for scanners to distinguish if they can read the "format" of the resource that is located at that URL or not (seems very similar to a Media Type now that I am writing it out).

## Alternatives

I am aware of the work being done on the Transparancy Exchange API. To me the TEA seems like a parallel usecase but not contradictory. E.g. maybe the linked VEX file might even be hosted by a TEA in the future.

**Comment by jkowalleck** [comment]
> We are already distributing SBOMs with our container images as In-Toto attestations and are very happy that scanners like Trivy are able to pick up the BOM and use it for the most accurate scan results we ever had.

So you have a static SBOM and a dynamic VEX?
If so, then here is my solution:
- keep that SBOM static
- create a VEX periodically, and have the affected components [BOM-linked](https://cyclonedx.org/capabilities/bomlink/) via `$.vulnerabilities[].affects[].ref`.
Here is an appropriate example for you> <https://github.com/CycloneDX/bom-examples/blob/master/VEX/CISA-Use-Cases/Case-7> 

If you wanted to communicate the URL to the latest VEX in an SBOM, you could use an external-reference of `type` = `exploitability-statement`. Of course, you may update the content of that VEX periodically, while keeping the URL static.
https://github.com/CycloneDX/specification/blob/268138039498a4f390ddc5d7e7e78857a951986f/schema/bom-1.6.schema.json#L1808
Ingesting tools could check whether the external-reference's target is a CycloneDX document, by schema-validating it. This way they could distinguish CycloneDX VEX/VDR from PDF or other data.

**Comment by RingoDev** [comment]
Yes that would have been my proposal as well, thanks for the additional references!

The question that is a bit harder to answer is: do you see this being a use case other organizations will follow on as well so it makes sense for SBOM vulnerability scanners to take on the maintenance to support it? 

Why is that so important to us?

Realistically nobody will manually parse through SBOMs and check externalReferences for VEX information. In order to be useful, it needs to be so deeply ingrained in automatic processes and supported by the scanning tools directly.

So this is more coming from the direction -> is this the use case the spec expects to be built upon this externalReference->exploitability-statement field so we can go and rally some of the known SBOM scanning tools to support it as another way to discover VEX information?

**Comment by ppkarwasz** [comment]
> The question that is a bit harder to answer is: do you see this being a use case other organizations will follow on as well so it makes sense for SBOM vulnerability scanners to take on the maintenance to support it?

At Apache Logging, we have been adding a [link to our VDR](https://logging.apache.org/cyclonedx/vdr.xml) to each published SBOM since last year. Looking at our download stats, it doesn't seem that many scanners follow that link. Maybe for VEX-es the situation will be different.

IMHO those links will never be widely adopted, but they might be useful to populate an initial [Transparency Exchange API](https://github.com/CycloneDX/transparency-exchange-api) server.

**Comment by RingoDev** [comment]
Hey Piotr, is that what was also shortly discussed at KoalaCon?

The main difference here in my view is that scanners already rely on some very thorough sources for Vulnerability disclosure information (VDR) such as osv.dev or the NVD directly. So they might not have incentives to look for additional sources as long as they have the data already from somewhere else.

It may even incur additional overhead due to the fact that it would need to be de-duplicated against those central sources in many cases. Also a year does not seem that long ago yet, considering the current state of SBOM adoption😉

With VEX this still seems like a still very active point of discussion for scanners of where to discover such information where I expect Vendors probably want to have direct control over. 

Also I strongly believe that Vendors want to have automatic ways of communicating the fact that they are "not affected" by something and if they are able to link to such a source from an SBOM and scanners pick it up, that may even benefit SBOM adoption.

**Comment by ppkarwasz** [comment]
> Hey Piotr, is that what was also shortly discussed at KoalaCon?

Yes, I must have mentioned the fact.

> With VEX this still seems like a still very active point of discussion for scanners of where to discover such information where I expect vendors probably want to have direct control over (due to possible liability concerns).

Yes, VEX-es are different, since there is not centralized database for them yet. For VDR-s there is also the possibility that the original version has more information that what ends up in NVD.

> Also I strongly believe that vendors want to have automatic ways of communicating the fact that they are "not affected" by something and if they are able to link to such a source from an SBOM and scanners pick it up, that may even benefit SBOM adoption.

I certainly believe that putting links to VEX-es in SBOMs can be interesting. The big question is if this can be implemented on the entire supply chain. Right now it is easier for me, as a library developer, to issue VEX-es that state "CVE-12345 is probably exploitable, but `libfoo` 1.2.3 works well with the patched version 2.3.4 of the vulnerable `libbar`". It takes a CI run and 5 minutes to do that, no need to plunge into the code. :wink: Nevertheless I would be interested in establishing a test chain of dependent project, where each project publishes an SBOM and a real VEX.

**Comment by RingoDev** [comment]
Hey, to not leave that thread hanging: Trivy is now supporting the resolution of VEX documents linked via an externalReference when scanning CycloneDX SBOMs and uses them to filter scan results: https://github.com/aquasecurity/trivy/discussions/8495

I still believe it could be beneficial to further scope the definition of `externalReference` to include an optional `mediaType` declaration to make it easier for tools to decide if they should try to retrieve the linked resource. Maybe that could be a consideration for the future @jkowalleck ?

**Comment by elear** [comment]
Can you guys pop this one up on your list of priorities for 2.0?  Tools shouldn't have to use heuristics to figure out the syntax of a reference is.  Moreover, without the information, you cannot formally validate multiple files, because although they are linked, nothing in the spec states how to validate across those links.

**Comment by jkowalleck** [comment]
moving this to milestone 2.0 - maybe we have time for that. 
and at least we should not make things worse. 
CC @stevespringett

**Comment by elear** [comment]
@jkowalleck thank you!  As @ppkarwasz wrote, the easiest way to do this is to add, at least as optional, a media-type field, and some narrative text that states that for CycloneDX, the files are to be parsed as a group.  I will note that this information can also SOMETIMES be plucked off of the HTTP connection.  But not always.

**Comment by stevespringett** [comment]
IMO, the mediaType would help dramatically.

For example:
```
application/vnd.cyclonedx+json; version=2.0;
```

but that really doesn't help with formats that do not have a registered media type, such as CSAF. In another proposal (#185), the idea was floated to include support for the schema URL used to validate the external reference. In that case:

```
https://docs.oasis-open.org/csaf/csaf/v2.0/csaf_json_schema.json
```
would inform the consumer that the referenced VEX is a CSAF URL.

Ideally, both mediaType and schema could be used - either together or individually - to help consumers identify what's on the other end of the URL.

Thoughts?

**Comment by elear** [comment]
The media type is **easy** to do.  There's a [form](https://www.iana.org/form/media-types).  Not that it matters that much, but you could probably even get this one into the standards part of the tree.  The only bit of work is writing out a security considerations aspect.  That's made a little more difficult – albeit not impossible – through the use of *detached* signatures.
